### PR TITLE
Decimal Spiral

### PR DESCRIPTION
**Pull request by thesabinelim**
Added the `!decimalspiral` command to generate and print a decimal spiral of size defined by the user
- Usage: `!decimalspiral <size>` or `!ds <size>`
- `<size>` must be an odd integer between 5 to 43

### CODE REVIEW

File: commands/decimal_spiral.go [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+func (d *DecimalSpiral) Chans() []string {
+	return nil
+}
+
+func (d *DecimalSpiral) MsgHandle(ses *discordgo.Session, msg *discordgo.Message, args []string) (*CommandSend, error) {
+	if len(args) == 0 {
+		return nil, ErrDecimalSpiralArgs
+	}
+
+    size, err := strconv.Atoi(args[0]);
```
Comment (VineBalloon) [inline commands/decimal_spiral.go]: Go style mandates tabs, do not indent with spaces. I think you can setup gofmt on save in VSCode, something to look into.

File: commands/decimal_spiral.go [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+    if err != nil {
+        if err.(*strconv.NumError).Err == strconv.ErrSyntax {
+            return nil, ErrDecimalSpiralSyntax
+        } else if err.(*strconv.NumError).Err == strconv.ErrRange {
+            return nil, ErrDecimalSpiralRange
+        } else {
```
Comment (VineBalloon) [inline commands/decimal_spiral.go]: Don't need else on early exit conditions, avoid them as they add clutter.

File: commands/decimal_spiral.go [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+            return nil, err
+        }
+    }
+
+	if isEven(size) || size < DecimalSpiralLowerLimit || size > DecimalSpiralUpperLimit {
```
Comment (VineBalloon) [inline commands/decimal_spiral.go]: Avoid making functions that could be put on one line. We teach you to make simple functions in 1511 to help grasp the concept of abstractions. However, low-level abstractions like this only add calling overhead (setting up and taking down a stack frame for the function).
```suggestion
	if size % 2 == 0 || size < DecimalSpiralLowerLimit || size > DecimalSpiralUpperLimit {
```

File: commands/decimal_spiral.go [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+// genDecimalSpiral consists of 2 steps:
+// 1. Loop left to right, up to down, calling isDigit to determine
+// whether to print a digit or a dash for each coordinate
+// 2. If a digit should be printed, call getDigit to calculate
+// the digit to be printed for that coordinate
+// Returns the decimal spiral as a string
+func genDecimalSpiral(size int) string {
+	out := ""
+	row := 0
+    for row < size {
```
Comment (VineBalloon) [inline commands/decimal_spiral.go]: Spaces again.

File: commands/decimal_spiral.go [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+	}
+
+	return NewSimpleSend(msg.ChannelID, utils.Block(genDecimalSpiral(size))), nil
+}
+
+func isEven(i int) bool {
+	return i % 2 == 0
+}
+
+func abs(i int) int {
```
Comment (VineBalloon) [inline commands/decimal_spiral.go]: Spaces again.

File: commands/decimal_spiral.go [hunk 6]
```diff
@@ -0,0 +1,258 @@
...
+// the digit to be printed for that coordinate
+// Returns the decimal spiral as a string
+func genDecimalSpiral(size int) string {
+	out := ""
+	row := 0
+    for row < size {
+        col := 0
+        for col < size {
+            if isDigit(size, row, col) {
+				out += strconv.Itoa(getDigit(size, row, col) % 10)
```
Comment (VineBalloon) [inline commands/decimal_spiral.go]: Indentation issues.

File: commands/decimal_spiral.go [hunk 7]
```diff
@@ -0,0 +1,258 @@
...
+}
+
+// isDigit determines whether a coordinate on a square is part of
+// the spiral of digits and returns true if it is, and false if not
+func isDigit(size int, row int, col int) bool {
+    // Absolute row distance from midpoint
+    rowDist := abs(row - size / 2)
+    // Absolute column distance from midpoint
+    colDist := abs(col - size / 2)
+	isDigit := false
```
Comment (VineBalloon) [inline commands/decimal_spiral.go]: Indentation issues.

File: commands/decimal_spiral.go [hunk 8]
```diff
@@ -0,0 +1,258 @@
...
+        row++
+    }
+
+    return out
+}
+
+// isDigit determines whether a coordinate on a square is part of
+// the spiral of digits and returns true if it is, and false if not
+func isDigit(size int, row int, col int) bool {
+    // Absolute row distance from midpoint
```
Comment (VineBalloon) [inline commands/decimal_spiral.go]: Spaces again.

File: commands/decimal_spiral.go [hunk 9]
```diff
@@ -0,0 +1,258 @@
...
+// accordingly to determine how much to add or subtract from the
+// corner values
+func getDigit(size int, row int, col int) int {
+	// Special handling for box segments modified to be spirals
+	if (row <= size / 2 && col < size / 2 && row == col + 1) {
+		row = row -1
+		col = col - 1
+	}
+
+    // Row displacement from midpoint
```
Comment (VineBalloon) [inline commands/decimal_spiral.go]: Spaces again.

File: commands/decimal_spiral.go (hunk 3 above)
Comment (thesabinelim) [reply commands/decimal_spiral.go]: Oh, thank you for this tip!

**Pull request merged**